Adds an ~5 minute guide to Numba.

stuartarchibald opened the pull request:
As title.

<!--

Thanks for wanting to contribute to Numba :)

First, if you need some help or want to chat to the core developers, please
visit https://gitter.im/numba/numba for real time chat or post to the Numba
mailing list https://groups.google.com/a/continuum.io/forum/#!forum/numba-users.

Here's some guidelines to help the review process go smoothly.

0. Please write a description in this text box of the changes that are being
   made.

1. Please ensure that you have written units tests for the changes made/features
   added.

2. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

3. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]` and
   then remove the label when you'd like it to be reviewed.

4. Once review has taken place please do not add features or make changes out of
   the scope of those requested by the reviewer (doing this just add delays as
   already reviewed code ends up having to be re-reviewed/it is hard to tell
   what is new etc!). Further, please do not rebase your branch on master/force
   push/rewrite history, doing any of these causes the context of any comments
   made by reviewers to be lost. If conflicts occur against master they should
   be resolved by merging master into the branch used for making the pull
   request.

Many thanks in advance for your cooperation!

-->

stuartarchibald commented:
This is a WIP draft of a 5 minute guide to Numba. I'm not sure where it is going just yet, I've shoved it into the toctree for the documentation and written it in RST just for convenience so we can get a rendered version to look at. This will need adjusting if it does go into the formal documentation.

This guide is written in more colloquial style than the rest of the documentation (that is typically formal).  It is intended for new users/those who've just heard of Numba and want to see if it is a likely to be a good fit for them. It therefore focuses on how to get set up to use Numba, what Numba is good at and how to correctly assess whether it is doing a good job.

Feedback is very welcome.

TODO: 
- [ ] Fix up links based on whether this goes in docs or elsewhere.

seibert commented on docs/source/user/5minguide.rst:
```diff
@@ -0,0 +1,247 @@
+
+  $ pip install numba
+
+Numba can also be compiled from source (`instructions <LINK>`_)
```
", although we do not recommend it for first-time Numba users."

seibert commented on docs/source/user/5minguide.rst:
```diff
@@ -0,0 +1,247 @@
+  messages
+* ``pyyaml`` - enables configuration of Numba via a YAML config file.
+* ``icc_rt`` - allows the use of the Intel SVML (high performance short vector
+  math library, x86_64 only).
```
Link to performance guide on icc_rt, so people see how to install it.

seibert commented on docs/source/user/5minguide.rst:
```diff
@@ -0,0 +1,247 @@
+
+    x = np.arange(100).reshape(10, 10)
+
+    @jit
```
Although you discuss nopython mode below, I would *start* the example with nopython mode and then rephrase the section below to explain why you do this.  I'd rather show best general practices, and then explain when to break them, than vice versa.

seibert commented on docs/source/user/5minguide.rst:
```diff
@@ -0,0 +1,247 @@
+            acc +=i
+        return a, acc
+
+    print(loop_lift(x))
```
At this point, I consider loop lifting to be an advanced Numba feature that we can leave out of a quickstart guide.

seibert commented on docs/source/user/5minguide.rst:
```diff
@@ -0,0 +1,247 @@
+run this code via the interpreter but with the added cost of the Numba internal
+overheads!
+
+The ``nopython`` mode *gotcha*.
```
Retitle this as "What is ``nopython`` mode?"

seibert commented on docs/source/user/5minguide.rst:
```diff
@@ -0,0 +1,247 @@
+
+The ``nopython`` mode *gotcha*.
+-------------------------------
+Pretty much all first time users are caught by this so it deserves a mention...
```
I would shorten this section and write it assuming nopython mode is the standard way to use `@jit`, mention briefly that Numba has limited support for mixed mode via loop lifting and maybe link to somewhere in the docs that give more information.

seibert commented on docs/source/user/5minguide.rst:
```diff
@@ -0,0 +1,247 @@
+    x = np.arange(100).reshape(10, 10)
+
+    @jit
+    def go_fast(a): # Function is compiled and runs in machine code
```
Change this comment to `# Function is compiled to machine code when called the first time`

seibert commented on docs/source/user/5minguide.rst:
```diff
@@ -0,0 +1,247 @@
+A good way to measure the impact Numba JIT has on your code is to time execution
+using the `timeit <https://docs.python.org/3/library/timeit.html>`_ module
+functions, these measure multiple iterations of execution and, as a result,
+can be made to accommodate for the compilation time in the first execution.
```
Show example of doing `%timeit` in the Jupyter Notebook / IPython as well.

seibert commented on docs/source/user/5minguide.rst:
```diff
@@ -0,0 +1,247 @@
+
+How does Numba work?
+--------------------
+Glossing over a huge amount of detail, Numba reads the Python bytecode for a
```
Don't need "Glossing over a huge amount of detail".  😄

seibert commented on docs/source/user/5minguide.rst:
```diff
@@ -0,0 +1,247 @@
+Glossing over a huge amount of detail, Numba reads the Python bytecode for a
+decorated function and combines this with information about the types of the
+input arguments to the function. It then forms a Numba specific internal
+representation(IR) of the function and translates this to LLVM IR, it then
```
We can go even simpler on this discussion.  "It analyzes and optimizes your code, and finally uses the LLVM compiler library to generate a machine code version of your function, tailored to your CPU capabilities."

seibert commented on docs/source/user/5minguide.rst:
```diff
@@ -0,0 +1,247 @@
+
+ctypes/cffi/cython interoperability:
+
+* ``cffi`` - `CFFI  <LINK NB CFFI DOCS>`_ functions are supported.
```
I would phrase these as "Calling X functions is supported in nopython mode."

seibert commented:
I think we want to pair this with a notebook of code examples, and then embed a link near the top of the guide to https://mybinder.org/ so they can try them out in their browser.

stuartarchibald commented:
Thanks for the review @seibert. I'm happy for this to be superseded by the outcome of the front-page description discussion on gitter, or for some amalgamation to happen if that's convenient, or this can stand as is? Thoughts?

stuartarchibald commented:
The only remaining broken link is to ROC which is in PR https://github.com/numba/numba/pull/3178. Should the example code/notebooks etc be done in a separate PR to this one?

seibert commented:
Examples and code can be a separate PR.

stuartarchibald commented:
ok, I think that's a good separation of work. I think I've addressed all the comments, so pending merge of ROC docs so I can update that link (and any other further review comments) I think this is sufficiently complete to consider merge :)

seibert commented:
ROC docs are merged now.

seibert commented:
Ready to merge as soon as one of the Travis CI build_doc configurations pass...

seibert commented:
Doc build passed.  Merging
